keps/0031-namespace-management.md

```diff
@@ -0,0 +1,167 @@
+authors:
+  - "@kensipe"
+owners:
+  - "@kensipe"
+editor: @kensipe
+creation-date: 2020-05-06
+last-updated: 2020-05-11
+status: draft
+---
+
+# Namespace Management
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Namespace Management in the Eco-System](#namespace-management-in-the-eco-system)
+    * [Kubernetes Objects and Namespaces](#kubernetes-objects-and-namespaces)
+    * [Helm](#helm)
+    * [The Many Meanings of a Namespace](#the-many-meanings-of-a-namespace)
+* [Proposals](#proposals)
+    * [Proposal 1: No namespace creation](#proposal-1-no-namespace-creation)
+    * [Proposal 2: Namespace Creation](#proposal-2-namespace-creation)
+      * [Alternative Ideas](#alternative-ideas)
+    * [Proposal: Multi-Namespace (static)](#proposal-multi-namespace-static)
+    * [Proposal: Dependencies and Creating Namespaces](#proposal-dependencies-and-creating-namespaces)
+* [Notes](#notes)
+* [Alternatives](#alternatives)
+
+
+## Summary
+
+This KEP aims to provide a definition on namespace management and provide guidance and constraints a KUDO operator may use namespaces.  This includes the expectations of creation, deletion, defining how an operator will install to a namespace and the ability of an operator to leverage more than one namespace.
+
+## Motivation
+
+Like many Kubernetes Object, there is an expectation that a KUDO operator is organized by namespace.  This KEP defines expectations around namespace creation and use.
+
+### Goals
+
+* Define if a namespace is a prerequistie to operator installation
+* Define if a namespace is created by KUDO
+* Define support for advance namespaces (namespaces with metadata)
+* Define if an operator can be installed into a new namespace
+* Define if an operator can leverage more than one namespace and under what conditions
+
+
+### Non-Goals
+
+
+## Namespace Management in the Eco-System
+
+### Kubernetes Objects and Namespaces
+
+When working kubernetes objects, be it Pods, ReplicaSets, Deployments and the like that require a namespace, the responsbility falls on the user to create that namespace prior assigning the resource or it will fail.  Available is an [auto-provision admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#namespaceautoprovision) available to auto-create namespaces. Users desiring auto-provisioning of namespaces should install this tool.  This webhook has the limitation of creating namespaces without extra metadata which defeats the value for use cases which need that metadata.
+
+### Helm
+
+Helm v1 and v2 provided an auto-provision capability.  The new Helm v3 (based on previous experiences) removed this capability and now requires users to create namespaces prior to installing a helm chart. Much detail is written in comments for [helm issue 6794](https://github.com/helm/helm/issues/6794)
```

Comment:
nits:

Should we mention Helm v3 supports `--create-namespace` if the user specifies it explicitly?
https://github.com/helm/helm/issues/6794